latok/core/latok_utils.py

```diff
@@ -114,3 +119,117 @@ def feature_weights(self):
         Get the non-zero feature names mapped to their weights.
         '''
         return {FEATURE_NAMES[idx]: self.features[idx] for idx in np.nonzero(self.features)[0]}
+
+    def add_abstract_feature(self, feature_name):
+        if self.abstract_features is None:
+            self.abstract_features = list()
+        self.abstract_features.append(feature_name)
+
+
+def _apply_combo_matrix(combo_matrix, mt):
+    s = _combine_matrix_rows(mt, combo_matrix)
+    return np.nonzero(s)[0]
+
+
+class OffsetSpec:
+    '''
+    Specification of offsets to be present and/or absent.
+    * present: list of offsets where features are to be present
+    * absent: list of offsets where features are to be absent
+    * pcmat: present combo matrix ndarray
+    * acmat: absent combo matrix ndarray
+    '''
+    def __init__(self, present=None, absent=None):
+        self.present = present
+        self.absent = absent
+        self.pcmat = None
+        self.acmat = None
+        if present is not None and len(present) > 0:
+            self.pcmat = build_combo_matrix([present])
+        if absent is not None and len(absent) > 0:
+            self.acmat = build_combo_matrix([absent])
+
+    @property
+    def has_present(self):
+        return self.present is not None and len(self.present) > 0
+
+    @property
+    def has_absent(self):
+        return self.absent is not None and len(self.absent) > 0
+
+    def matches(self, mt, pa_align=False):
+        result, p, a = False, None, None
+        if self.has_present:
+            p = _apply_combo_matrix(self.pcmat, mt)
+            if len(p) == 0:
+                return False
+            else:
+                result = True
+        if self.has_absent:
+            a = _apply_combo_matrix(self.acmat, mt)
+            if pa_align and p is not None:
+                a *= p
+            if len(a) > 0:
+                return False
+            else:
+                result = True
+        return result
+
+
+@dataclass
+class FeatureSpec:
+    '''
```

Comment:
Also not sure I understand purpose of this class.

Plus minor concern about using regex. Rather than use regex, can this be folded into the initial vectorization? Seems like it would slow things down to iteratively apply regex to the same string we just vectorized

Reply:
This is intended as an initial attempt to show how abstract featurization can be incorporated into the tokenizer and to exercise it concretely. Once any kinks are worked out of this process, it can be refactored into the more performant paradigm we've started with the c-extensions. In fact, a path for this is already starting to appear (in my mind) for doing this by associating indexes of matrix row masks and where they apply in a string for quick retrieval of this information instead of its reapplication as is currently happening.

As for adding allowance for regular expressions, it is present here for the purpose of exploring additional mechanisms that best provide for abstract featurization. My intuition is that the regular expressions will likely not be necessary at this level in the long run, but I wanted to keep the option open during exploration.

I'm good to either continue the evolution of these concepts in this branch before merging to master or to tie off this branch and do that refactoring work over time. Whatever you prefer.